securedrop/journalist_app/sessions.py

```diff
@@ -0,0 +1,182 @@
+from typing import Optional, Any
+from flask.sessions import SessionInterface as FlaskSessionInterface
+from flask.sessions import SessionMixin, session_json_serializer
+from json.decoder import JSONDecodeError
+from redis import Redis
+from werkzeug.datastructures import CallbackDict
+from itsdangerous import URLSafeTimedSerializer, BadSignature
+
+
+class ServerSideSession(CallbackDict, SessionMixin):
+    """Baseclass for server-side based sessions."""
+
+    def __init__(self, sid: 'str', initial: 'Any' = None) -> 'None':
+        def on_update(self: ServerSideSession) -> None:
+            self.modified = True
+        CallbackDict.__init__(self, initial, on_update)
+        self.sid = sid
+        self.token = None
+        self.is_api = False
+        self.to_destroy = False
+        self.to_regenerate = False
+        self.modified = False
+
+
+class RedisSession(ServerSideSession):
+    def get_token(self) -> Optional['str']:
+        return self.token
+
+    def get_lifetime(self) -> 'datetime':
+        return datetime.utcnow() + timedelta(seconds=2 * 60 * 60)
+
+    def destroy(self) -> 'None':
+        self.to_destroy = True
+
+    def regenerate(self) -> 'None':
+        self.to_regenerate = True
+
+
+class SessionInterface(FlaskSessionInterface):
+
+    def _generate_sid(self) -> 'str':
+        return token_urlsafe(32)
+
+    def _get_signer(self, app: 'Flask') -> 'URLSafeTimedSerializer':
+        return URLSafeTimedSerializer(app.secret_key,
+                                      salt=self.salt)
+
+
+class RedisSessionInterface(SessionInterface):
+    """Uses the Redis key-value store as a session backend.
+
+    :param redis: A ``redis.Redis`` instance.
+    :param key_prefix: A prefix that is added to all Redis store keys.
+    :param salt: Allows to set the signer salt from the calling interface
+    :param header_name: if use_header, set the header name to parse
+    """
+
+    session_class = RedisSession
+
+    def __init__(self, lifetime: 'int', renew_count: 'int', redis: 'Optional[Redis]',
+                 key_prefix: 'str', salt: 'str', header_name: 'str') -> 'None':
+        self.serializer = session_json_serializer
+        if redis is None:
+            redis = Redis()
+        self.redis = redis
+        self.lifetime = lifetime
+        self.renew_count = renew_count
+        self.key_prefix = key_prefix
+        self.api_key_prefix = 'api_' + key_prefix
+        self.salt = salt
+        self.api_salt = 'api_' + salt
+        self.header_name = header_name
+        self.new = False
+        self.has_same_site_capability = hasattr(self, "get_cookie_samesite")
+
+    def open_session(self, app: 'Flask', request: 'Request') -> 'Optional[SessionMixin]':
+        '''This function is called by the flask session interface at the
+        beginning of each request.
+        '''
+        is_api = (request.path.split('/')[1] == 'api')
+
+        def new_session(is_api: 'bool' = False) -> 'SessionMixin':
```

Comment:
Why not make this a normal function on the RedisSessionInterface class? If it should be private, just prefix it with an underscore.